enhancements/kube-apiserver/api-webhook-supportability.md

```diff
@@ -0,0 +1,341 @@
+---
+title: Improving Supportability of API Webhooks
+authors:
+- "@sttts"
+  reviewers:
+- "@mfojtik"
+- "@tkashem"
+- "@p0lynomial"
+- "@deads2k"
+  approvers:
+- "@mfojtik"
+  creation-date: 2021-09-03
+  last-updated: 2021-09-23
+  status: implementable
+  see-also: []
+  replaces: []
+  superseded-by: []
+---
+
+# Improving Supportability of API webhooks
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Operational readiness criteria is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+OpenShift supports the API extensions that upstream Kubernetes has built into the kube-apiserver. This especially involves webhooks for admission and for CRD conversion. This enhancement is about measures to increase supportability of clusters that use official Red Hat provided add-ons (like Istio, kubevirt or Open Policy Agent), but also 3rd-party add-ons (like Vault, or the community variants of previously mentioned software).
+
+API extension webhooks can have negative impact on a cluster, from performance, over stability and security to availability. Customer case escalations of early adopters have shown that these concerns are very real and supportability is at risk.
+
+A formal tainting mechanism with influence on the support status of a cluster is not desired by the OpenShift organization.
+
+Hence, we propose:
+- a number of **alerts** to warn about unreliability of webhooks or performance problems
+- a number of **informational alerts** notifying the admin about 3rd-party software that installs webhooks on critical resources
+- a non-centralized "soft" **approval process** (similar to the use of `k8s.io` group names in CRDs in upstream) for webhooks hosted in `openshift-*` namespaces for core resources
+- an **extension of OpenShift enhancement template** about webhooks, their impact to the cluster, possible debugging strategies and supportability concerns
+- a **monitoring dashboard** to watch the latency of individual webhooks and total latency for all webhooks per resource.
+- to **skip 3rd-party admission webhooks on runlevel=0 namespaces**
```

Comment:
missing additional detail on this.  I expect this to be a controversial claim and it will need a process that allows bypassing it. Sometimes "all" really means "all".

Reply:
Will add. tl/dr: carry patch to skip the namespace in the webhook admission plugin.

Reply:
done